internal/cephfs/controllerserver.go

```diff
@@ -298,3 +410,291 @@ func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi
+		return snap, err
+	}
+	var t *timestamp.Timestamp
+	t, err = parseTime(ctx, snap.CreatedAt)
+	if err != nil {
+		return snap, err
+	}
+	snap.CreationTime = t
+	err = protectSnapshot(ctx, volOpt, cr, snapID, volID)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to protect snapshot %s %v"), snapID, err)
+	}
+	return snap, err
+}
+
+func parseTime(ctx context.Context, createTime string) (*timestamp.Timestamp, error) {
+	tm := &timestamp.Timestamp{}
+	layout := "2006-01-02 15:04:05.000000"
+	// TODO currently parsing of timestamp to time.ANSIC generate from ceph fs is failng
+	var t time.Time
+	t, err := time.Parse(layout, createTime)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to parse time %s %v"), createTime, err)
+		return tm, err
+	}
+	tm, err = ptypes.TimestampProto(t)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to convert time %s %v"), createTime, err)
+		return tm, err
+	}
+	return tm, nil
+}
+func (cs *ControllerServer) validateSnapshotReq(ctx context.Context, req *csi.CreateSnapshotRequest) error {
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		klog.Errorf(util.Log(ctx, "invalid create snapshot req: %v"), protosanitizer.StripSecrets(req))
+		return err
+	}
+
+	// Check sanity of request Snapshot Name, Source Volume Id
+	if req.Name == "" {
+		return status.Error(codes.InvalidArgument, "snapshot Name cannot be empty")
+	}
+	if req.SourceVolumeId == "" {
+		return status.Error(codes.InvalidArgument, "source Volume ID cannot be empty")
+	}
+
+	return nil
+}
+
+// DeleteSnapshot deletes the snapshot in backend and removes the
+// snapshot metadata from store.
+func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		klog.Errorf(util.Log(ctx, "invalid delete snapshot req: %v"), protosanitizer.StripSecrets(req))
+		return nil, err
+	}
+
+	cr, err := util.NewAdminCredentials(req.GetSecrets())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+	defer cr.DeleteCredentials()
+	snapshotID := req.GetSnapshotId()
+	if snapshotID == "" {
+		return nil, status.Error(codes.InvalidArgument, "snapshot ID cannot be empty")
+	}
+
+	if acquired := cs.SnapshotLocks.TryAcquire(snapshotID); !acquired {
+		klog.Errorf(util.Log(ctx, util.SnapshotOperationAlreadyExistsFmt), snapshotID)
+		return nil, status.Errorf(codes.Aborted, util.SnapshotOperationAlreadyExistsFmt, snapshotID)
+	}
+	defer cs.SnapshotLocks.Release(snapshotID)
+
+	volOpt, snapInfo, sid, err := newSnapshotOptionsFromID(ctx, snapshotID, cr)
+	if err != nil {
+		// if error is ErrPoolNotFound, the pool is already deleted we dont
+		// need to worry about deleting snapshot or omap data, return success
+		var epnf util.ErrPoolNotFound
```

Comment:
I think these are now sentinel errors, and can be checked with `errors.Is()`

Reply:
yep. done.